internal/hscan/hscan_test.go

```diff
@@ -0,0 +1,117 @@
+package hscan
+
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+type data struct {
+	Omit  string `redis:"-"`
+	Empty string
+
+	String string  `redis:"string"`
+	Bytes  []byte  `redis:"byte"`
+	Int    int     `redis:"int"`
+	Uint   uint    `redis:"uint"`
+	Float  float32 `redis:"float"`
+	Bool   bool    `redis:"bool"`
+}
+
+type i []interface{}
+
+func TestGinkgoSuite(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "hscan")
+}
+
+var _ = Describe("Scan", func() {
+	It("catches bad args", func() {
+		var d data
+
+		Expect(Scan(&d, i{}, i{})).NotTo(HaveOccurred())
+		Expect(d).To(Equal(data{}))
+
+		Expect(Scan(&d, i{"key"}, i{})).To(HaveOccurred())
+		Expect(Scan(&d, i{"key"}, i{"1", "2"})).To(HaveOccurred())
+		Expect(Scan(nil, i{"key", "1"}, i{})).To(HaveOccurred())
+
+		var m map[string]interface{}
+		Expect(Scan(&m, i{"key"}, i{"1"})).To(HaveOccurred())
+		Expect(Scan(data{}, i{"key"}, i{"1"})).To(HaveOccurred())
+		Expect(Scan(data{}, i{"key", "string"}, i{nil, nil})).To(HaveOccurred())
+	})
+
+	It("scans good values", func() {
+		var d data
+
+		// non-tagged fields.
+		Expect(Scan(&d, i{"key"}, i{"value"})).NotTo(HaveOccurred())
+		Expect(d).To(Equal(data{}))
+
+		keys := i{"string", "byte", "int", "uint", "float", "bool"}
+		vals := i{"str!", "bytes!", "123", "456", "123.456", "1"}
+		Expect(Scan(&d, keys, vals)).NotTo(HaveOccurred())
+		Expect(d).To(Equal(data{
+			String: "str!",
+			Bytes:  []byte("bytes!"),
+			Int:    123,
+			Uint:   456,
+			Float:  123.456,
+			Bool:   true,
+		}))
+
+		// Scan a different type with the same values to test that
+		// the struct spec maps don't conflict.
+		type data2 struct {
+			String string  `redis:"string"`
+			Bytes  []byte  `redis:"byte"`
+			Int    int     `redis:"int"`
+			Uint   uint    `redis:"uint"`
+			Float  float32 `redis:"float"`
+			Bool   bool    `redis:"bool"`
+		}
+		var d2 data2
```

Comment:
Just FYI this can be written as

```go
var d2 struct {
			String string  `redis:"string"`
			Bytes  []byte  `redis:"byte"`
			Int    int     `redis:"int"`
			Uint   uint    `redis:"uint"`
			Float  float32 `redis:"float"`
			Bool   bool    `redis:"bool"`
}
```